### PR TITLE
global: handle files between records and workflows

### LINKED ISSUE #2551: WF: plots
Plots and attach to metadata

### LINKED ISSUE #2552: WF: PDFs
Workflows: PDFs and attach to metadata

### PR DESCRIPTION
**Pull request by zzacharo**
## Description:
Adds the ``files`` property to ``InspireRecord`` by inheriting from the
``Record`` of ``invenio_records_files``. This allows us to attach plots
and documents to records (closes #2551 and closes #2552).

Also modifies those workflows tasks that deal with records so that they
use builder methods to populate the ``documents`` and ``figures`` keys,
and those will then be converted by DoJSON to the ``FFT`` fields that
Legacy understands.

## Related Issues:
Closes #2551 and closes #2552.

## Checklist:
- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [x] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [ ] My code follows the code style of this project.
- [ ] I've added any new docs if API/utils methods were added.
- [ ] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by zzacharo** [comment]
Missing integration tests. I checked locally with the database. I pushed it to start discussing about it :)

File: inspirehep/config.py [hunk 1]
```diff
@@ -1419,6 +1419,11 @@ def _(x):
     'physics.acc-ph',
     'physics.ins-det',
 ]
+RECORDS_DEFAULT_FILE_LOCATION_NAME = "records"
+"""Name of default records Location reference."""
+
+RECORDS_DEFAULT_STORAGE_CLASS = "S"
+"""Default storage class for record files."""
```
Comment (jacquerie) [inline inspirehep/config.py]: I might be misremembering, but isn't this the default of `invenio-records-files`

File: inspirehep/modules/workflows/tasks/upload.py [hunk 2]
```diff
@@ -48,6 +49,10 @@ def store_record(obj, *args, **kwargs):
     # we're sure that the schema will be validated without touching the full
     # holdingpen stack.
     record = InspireRecord.create(obj.data, id_=None)
+    from celery.contrib import rdb; rdb.set_trace()
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/upload.py]: Travis is probably going to be quite offended by this...

Comment (zzacharo) [inline inspirehep/modules/workflows/tasks/upload.py]: oups..

File: inspirehep/config.py (hunk 1 above)
Comment (zzacharo) [inline inspirehep/config.py]: Yes I think so but it was explicitly declared by workflows so I just used it.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 3]
```diff
@@ -133,12 +135,22 @@ def arxiv_plot_extract(obj, eng):
             current_app.logger.exception(err)
             return
 
+        lb = LiteratureBuilder(source='arxiv', record=obj.data)
```
Comment (zzacharo) [inline inspirehep/modules/workflows/tasks/arxiv.py]: @jacquerie  @david-caro What should I use for source here because the workflow object doesn't have at this point the `source` field set.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 4]
```diff
@@ -133,12 +135,22 @@ def arxiv_plot_extract(obj, eng):
             current_app.logger.exception(err)
             return
 
+        lb = LiteratureBuilder(source=obj.data['acquisition_source']['source'], record=obj.data)
         for idx, plot in enumerate(plots):
+            key = '{uuid}_{name}'.format(uuid=uuid.uuid4(), name=plot.get('name'))
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/arxiv.py]: do we really need a UUID here? it seems a bit overkill and will result in very long file names.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 5]
```diff
@@ -133,12 +135,22 @@ def arxiv_plot_extract(obj, eng):
...
             with open(plot.get('url')) as plot_file:
-                obj.files[plot.get('name')] = plot_file
-            obj.files[plot.get('name')]['description'] = u'{0:05d} {1}'.format(
+                obj.files[key] = plot_file
+            obj.files[key]['description'] = u'{0:05d} {1}'.format(
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/arxiv.py]: there is no description key in [`_files`](http://inspire-schemas.readthedocs.io/en/latest/schemas/records/elements/records-files.html#records-files-json). You can remove this line.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 6]
```diff
@@ -133,12 +135,22 @@ def arxiv_plot_extract(obj, eng):
...
                 idx, ''.join(plot.get('captions', []))
             )
+            lb.add_figure(
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/arxiv.py]: `source='arXiv'`

**Comment by zzacharo** [comment]
@michamos @jacquerie Can you check again? I created another commit for the pdfs. Missing tests.

File: inspirehep/modules/workflows/tasks/actions.py [hunk 7]
```diff
@@ -207,6 +208,14 @@ def submission_fulltext_download(obj, eng):
         )
 
         if pdf:
+            lb = LiteratureBuilder(source=obj.data['acquisition_source']['source'], record=obj.data)
+            lb.add_document(
+                filename,
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/actions.py]: how do you ensure uniqueness here? maybe the builder should raise an exception if the key is already present.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 8]
```diff
@@ -133,12 +146,31 @@ def arxiv_plot_extract(obj, eng):
             current_app.logger.exception(err)
             return
 
+        lb = LiteratureBuilder(source=obj.data['acquisition_source']['source'], record=obj.data)
         for idx, plot in enumerate(plots):
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/arxiv.py]: you don't need the `enumerate` any more, as `idx` is not used.

File: inspirehep/modules/workflows/tasks/upload.py [hunk 9]
```diff
@@ -49,15 +50,32 @@ def store_record(obj, *args, **kwargs):
...
+    # Workflow save `_files` with workflow objects files information
+    # When setting the record.files the `_files` are not overridden
+    # because in `FilesIterator.flush` method ignores the `_files`
+    # update if exists one in the record.
+
+    for key in obj.files.keys:
+        with open(os.path.realpath(obj.files[key].file.uri)) as stream:
+            record.files[key] = stream
+            key_updated = False
+            for document in record.get('documents', []):
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/upload.py]: you can loop only once for `documents` and `figures` instead of doing the `key_updated` thing:
```
for attachment in itertools.chain(record.get('documents', []), record.get('figures', [])):
```

Comment (zzacharo) [inline inspirehep/modules/workflows/tasks/upload.py]: I added because I want to omit iterating to figures if I have already find the key. Probably there is a better way finding it.

File: inspirehep/modules/workflows/tasks/upload.py [hunk 10]
```diff
@@ -49,15 +52,24 @@ def store_record(obj, *args, **kwargs):
...
 
+    # Workflow save `_files` with workflow objects files information
+    # When setting the record.files the `_files` are not overridden
+    # because in `FilesIterator.flush` method ignores the `_files`
+    # update if exists one in the record.
+
+    for key in obj.files.keys:
+        with open(os.path.realpath(obj.files[key].file.uri)) as stream:
+            record.files[key] = stream
+            for attachment in itertools.chain(record.get('documents', []), record.get('figures', [])):
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/upload.py]: I think you should have left the test for `attachment['key'] == key`.

Comment (zzacharo) [inline inspirehep/modules/workflows/tasks/upload.py]: oops...Yes indeed!

**Comment by zzacharo** [comment]
@david-caro @jacquerie @michamos can you review this one?

File: inspirehep/bat/pages/create_literature.py [hunk 11]
```diff
@@ -454,6 +454,40 @@ def _submit_doi_id():
...
+    )
+
+    _skip_import_data()
+
+    Arsenic().find_element_by_xpath(
+        '//div[@id="webdeposit_form_accordion"]/div[4]/span/button'
+    ).click()
+    Arsenic().show_title_bar()
+
+    return ArsenicResponse(_import_and_submit_arxiv_by_id)
```
Comment (jacquerie) [inline inspirehep/bat/pages/create_literature.py]: Why is a PR that add supports for files doing this?

File: inspirehep/bat/pages/create_literature.py [hunk 12]
```diff
@@ -520,3 +554,10 @@ def _select_chapter():
     Select(Arsenic().find_element_by_id('type_of_doc')).select_by_value(
         'chapter'
     )
+
+
+def check_expected_data_in_record(expected_data, record):
+    check = True
+    for key in expected_data:
+        check = check and expected_data[key] in record
+    return check
```
Comment (jacquerie) [inline inspirehep/bat/pages/create_literature.py]: Or this?

File: inspirehep/bat/pages/create_literature.py (hunk 11 above)
Comment (zzacharo) [reply inspirehep/bat/pages/create_literature.py]: I know but this was a refactoring needed to support additional acceptance tests by reusing components. Everything was hardcoded before.

File: tests/unit/workflows/test_workflows_tasks_arxiv.py [hunk 13]
```diff
@@ -271,15 +271,6 @@ def test_arxiv_plot_extract_populates_files_with_plots(mock_os):
...
 
-        expected = obj.files['figure1']['description']
-        result = (
-            '00000 Difference (in MeV) between the theoretical and '
-            'experimental masses for the 2027 selected nuclei as a '
-            'function of the mass number.'
-        )
-
-        assert expected == result
-
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_arxiv.py]: Why are we losing this? Where is it going?

File: tests/unit/conftest.py [hunk 14]
```diff
@@ -52,6 +52,7 @@ def app():
         CELERY_RESULT_BACKEND='cache',
         CELERY_CACHE_BACKEND='memory',
         CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
+        MAIL_SUPPRESS_SEND=True,
```
Comment (jacquerie) [inline tests/unit/conftest.py]: Do we need this?

File: tests/integration/test_permissions.py [hunk 15]
```diff
@@ -157,6 +157,7 @@ def test_inspire_search_filter_public_collection(app, app_client, user_info, tot
     # Doing a client request creates a request context that allows the
     # assert to correctly use the logged in user.
     app_client.get('/search')
+
```
Comment (jacquerie) [inline tests/integration/test_permissions.py]: These unrelated changes are why we get conflicts...

File: tests/unit/conftest.py (hunk 14 above)
Comment (zzacharo) [reply tests/unit/conftest.py]: leftover.

File: tests/integration/test_permissions.py (hunk 15 above)
Comment (zzacharo) [reply tests/integration/test_permissions.py]: my conflicts were due to `modules.workflows.tasks.arxiv` refactoring.

File: tests/acceptance/literature/test_literature_new_form.py [hunk 16]
```diff
@@ -116,7 +121,39 @@ def test_literature_create_thesis_manually(login):
...
+    }
+
+    _check_back_office(expected_data)
+    holding_panel_literature_detail.go_to()
+    submission_number = top_navigation_page.current_url().split('/')[-1]
+
+    from invenio_records.api import RecordMetadata
+
+    record = [record for record in RecordMetadata.query.all()
+              if 'acquisition_source' in record.json and record.json['acquisition_source']['submission_number'] == submission_number][0]
```
Comment (jacquerie) [inline tests/acceptance/literature/test_literature_new_form.py]: So, in order to get a record with files you create a new acceptance test (which requires you to modify some existing code in the acceptance test suite) in which you submit a record through the UI, you click all the buttons you need to click, then you look which URL you get and finally you get this record by using the id you parsed from the URL.

Perhaps there is a simpler way?

Comment (zzacharo) [reply tests/acceptance/literature/test_literature_new_form.py]: We decided to do this in order not to generate the whole workflow process in order to submit a new record. We can discuss another way if there is something easier.

**Comment by zzacharo** [comment]
@jacquerie do you propose to test only the functionality if I just attach a file to a record then check if I can retrieve it back and omit the acceptance test testing the submission?

File: inspirehep/modules/workflows/tasks/actions.py [hunk 17]
```diff
@@ -37,6 +37,7 @@
 )
 from inspirehep.utils.record import get_arxiv_id
 from inspirehep.utils.url import is_pdf_link
+from inspire_schemas.builders import LiteratureBuilder
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/actions.py]: Nit: this guy should be at the beginning of the `inspire` imports (alphabetical order)

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 18]
```diff
@@ -40,6 +40,7 @@
 from inspire_dojson.hep import hep
 from inspire_schemas.utils import classify_field
 from inspirehep.modules.workflows.utils import convert
+from inspire_schemas.builders import LiteratureBuilder
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/arxiv.py]: Nit: same thing, between `inspire_dojson` and the other `inspire_schemas` one.

File: inspirehep/modules/workflows/tasks/arxiv.py [hunk 19]
```diff
@@ -96,6 +107,7 @@ def arxiv_package_download(obj, eng):
     :param obj: Workflow Object to process
     :param eng: Workflow Engine processing the object
     """
+
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/arxiv.py]: Nit: no need to add this extra space.

File: inspirehep/modules/workflows/tasks/arxiv.py (hunk 18 above)
Comment (zzacharo) [reply inspirehep/modules/workflows/tasks/arxiv.py]: changed.

File: inspirehep/modules/workflows/tasks/actions.py (hunk 17 above)
Comment (zzacharo) [reply inspirehep/modules/workflows/tasks/actions.py]: changed.

**Comment by zzacharo** [comment]
@jacquerie made the changes.
